File: spring-data-requery-java/src/main/java/com/coupang/springframework/data/requery/core/RequeryOperations.java
package com.coupang.springframework.data.requery.core;

import com.coupang.springframework.data.requery.mapping.RequeryMappingContext;
import com.coupang.springframework.data.requery.utils.EntityDataStoreUtils;
import io.requery.TransactionIsolation;
import io.requery.meta.Attribute;
import io.requery.meta.EntityModel;
import io.requery.meta.QueryAttribute;
import io.requery.query.*;
import io.requery.sql.EntityContext;
import io.requery.sql.EntityDataStore;
import org.jetbrains.annotations.Nullable;

import javax.validation.constraints.NotNull;
import java.util.concurrent.Callable;
import java.util.function.Function;

/**
 * Java용 RequeryOperations
 *
 * @author debop@coupang.com
 * @since 18. 6. 4
 */
public interface RequeryOperations {

    EntityDataStore<Object> getDataStore();

    RequeryMappingContext getMappingContext();

    default EntityModel getEntityModel() {
        return EntityDataStoreUtils.getEntityModel(getDataStore());
    }

    @SuppressWarnings("unchecked")
    default <E> EntityContext<E> getEntityContext() {
        return EntityDataStoreUtils.getEntityContext(getDataStore());
    }

    default <E> Selection<? extends Result<E>> select(Class<E> entityType) {
        return getDataStore().select(entityType);
    }

    default <E> Selection<? extends Result<E>> select(Class<E> entityType, QueryAttribute<?, ?>... attributes) {
        return getDataStore().select(entityType, attributes);
    }

    default Selection<? extends Result<Tuple>> select(Expression<?>... expressions) {
        return getDataStore().select(expressions);
    }

    default <E, K> E findById(Class<E> entityType, K id) {
        return getDataStore().findByKey(entityType, id);
    }

    default <E> Iterable<E> findAll(Class<E> entityType) {
        return getDataStore().select(entityType).get().toList();
    }

    default <E> E refresh(@NotNull E entity) {
        return getDataStore().refresh(entity);
    }

    default <E> E refresh(@NotNull E entity, Attribute<?, ?>... attributes) {
        return getDataStore().refresh(entity, attributes);
    }

    default <E> Iterable<E> refresh(@NotNull Iterable<E> entities, Attribute<?, ?>... attributes) {
        return getDataStore().refresh(entities, attributes);
    }

    default <E> E refreshAll(@NotNull E entity) {
        return getDataStore().refreshAll(entity);
    }

    default <E> E upsert(@NotNull E entity) {
        return getDataStore().upsert(entity);
    }

    default <E> Iterable<E> upsertAll(@NotNull Iterable<E> entities) {
        return getDataStore().upsert(entities);
    }

    default <E> E insert(@NotNull E entity) {
        return getDataStore().insert(entity);
    }

    default <E, K> K insert(@NotNull E entity, Class<K> keyClass) {
        return getDataStore().insert(entity, keyClass);
    }

    default <E> Insertion<? extends Result<Tuple>> insert(Class<E> entityType) {
        return getDataStore().insert(entityType);
    }

    default <E> InsertInto<? extends Result<Tuple>> insert(Class<E> entityType, QueryAttribute<E, ?>... attributes) {
        return getDataStore().insert(entityType, attributes);
    }

    default <E> Iterable<E> insertAll(@NotNull Iterable<E> entities) {
        return getDataStore().insert(entities);
    }

    default <E, K> Iterable<K> insertAll(@NotNull Iterable<E> entities, Class<K> keyClass) {
        return getDataStore().insert(entities, keyClass);
    }


    default Update<? extends Scalar<Integer>> update() {
        return getDataStore().update();
    }

    default <E> E update(@NotNull E entity) {
        return getDataStore().update(entity);
    }

    default <E> E update(@NotNull E entity, Attribute<?, ?>... attributes) {
        return getDataStore().update(entity, attributes);
    }

    default <E> Update<? extends Scalar<Integer>> update(Class<E> entityType) {
        return getDataStore().update(entityType);
    }

    default <E> Iterable<E> updateAll(@NotNull Iterable<E> entities) {
        return getDataStore().update(entities);
    }

    default <E> Deletion<? extends Scalar<Integer>> delete() {
        return getDataStore().delete();
    }

    default <E> Deletion<? extends Scalar<Integer>> delete(Class<E> entityType) {
        return getDataStore().delete(entityType);
    }

    default <E> void delete(E entity) {
        getDataStore().delete(entity);
    }

    default <E> void deleteAll(Iterable<E> entities) {
        getDataStore().delete(entities);
    }

    default <E> Integer deleteAll(Class<E> entityType) {
        return getDataStore().delete(entityType).get().value();
    }

    default <E> Selection<? extends Scalar<Integer>> count(Class<E> entityType) {
        return getDataStore().count(entityType);
    }

    default <E> Selection<? extends Scalar<Integer>> count(QueryAttribute<?, ?>... attributes) {
        return getDataStore().count(attributes);
    }

    default Result<Tuple> raw(String query, Object... parameters) {
        return getDataStore().raw(query, parameters);
    }

    default <E> Result<E> raw(Class<E> entityType, String query, Object... parameters) {
        return getDataStore().raw(entityType, query, parameters);
    }

    default <V> V runInTransaction(Callable<V> callable) {
        return runInTransaction(callable, null);
    }

    <V> V runInTransaction(Callable<V> callable, @Nullable TransactionIsolation isolation);

    default <V> V withTransaction(Function<EntityDataStore<Object>, V> block) {
        return withTransaction(block, null);
    }

    <V> V withTransaction(Function<EntityDataStore<Object>, V> block, @Nullable TransactionIsolation isolation);
}

File: spring-data-requery-kotlin/src/main/kotlin/com/coupang/springframework/data/requery/core/RequeryOperations.kt
package com.coupang.springframework.data.requery.core

import io.requery.TransactionIsolation
import io.requery.meta.Attribute
import io.requery.meta.QueryAttribute
import io.requery.query.*
import io.requery.sql.EntityDataStore

/**
 * RequeryOperations
 *
 * @author debop@coupang.com
 * @since 18. 5. 29
 */
interface RequeryOperations {

    // TODO: default methods 로 하면 안되나?

    val dataStore: EntityDataStore<Any>

    fun <T> select(entityType: Class<T>): Selection<out Result<T>>

    fun <T> select(entityType: Class<T>, vararg attributes: QueryAttribute<T, *>): Selection<out Result<T>>

    fun select(vararg expressions: Expression<*>): Selection<out Result<Tuple>>

    fun <T, ID> findById(entityType: Class<T>, id: ID): T?

    fun <T> findAll(entityType: Class<T>): Iterable<T>

    fun <T> refresh(entity: T): T

    fun <T> refresh(entity: T, vararg attributes: Attribute<*, *>): T

    fun <T> refresh(entities: Iterable<T>, vararg attributes: Attribute<*, *>): Iterable<T>

    fun <T> refreshAll(entity: T): T

    fun <T> upsert(entity: T): T

    fun <T> upsertAll(entities: Iterable<T>): Iterable<T>

    fun <T> insert(entity: T): T

    fun <T> insertAll(entities: Iterable<T>): Iterable<T>

    fun <T> insert(entityType: Class<T>): Insertion<out Result<Tuple>>

    fun <T> insert(entityType: Class<T>, vararg attributes: QueryAttribute<T, *>): InsertInto<out Result<Tuple>>

    fun <T> update(entity: T): T

    fun <T> update(entityType: Class<T>): Update<out Scalar<Int>>

    fun <T> updateAll(entities: Iterable<T>): Iterable<T>

    fun <T> delete(entity: T)

    fun <T> delete(entityType: Class<T>): Deletion<out Scalar<Int>>

    fun <T> deleteAll(entities: Iterable<T>)

    fun <T> deleteAll(entityType: Class<T>): Long

    fun <T> count(entityType: Class<T>): Selection<out Scalar<Int>>

    fun raw(query: String, vararg parameters: Any?): Result<out Tuple>

    fun <T> raw(entityType: Class<T>, query: String, vararg parameters: Any?): Result<out T>

    fun <T> runInTransaction(block: RequeryOperations.() -> T): T

    fun <T> runInTransaction(isolation: TransactionIsolation, block: RequeryOperations.() -> T): T

    fun <T> withDataStore(block: EntityDataStore<Any>.() -> T): T

}